### PR TITLE
RouteProgress conversion to Kotlin

### LINKED ISSUE #2194: Convert RouteProgress to Kotlin
RouteProgress conversion to Kotlin

### PR DESCRIPTION
**Pull request by olegzil**
## Description
Closes #2194 
This PR converts RouteProgress to Kotlin. Additionally it modifies the code that uses this class. Some tests are failing, however, because this conversion is complex and it will take a bit of time to review, the PR is open for review now. The PR will be updated with fixes for the various failing tests.

- [ ] I have added any issue links
- [ ] I have added all related labels (`bug`, `feature`, `new API(s)`, `SEMVER`, etc.)
- [ ] I have added the appropriate milestone and project boards

### Goal
Convert RouteProgress to Kotlin

## Testing

Please describe the manual tests that you ran to verify your changes

- [x ] I have tested locally (including `SNAPSHOT` upstream dependencies if needed) through testapp/demo app and run all activities to avoid regressions
- [ ] I have tested via a test drive, or a simulation/mock location app
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Checklist

- [ x] My code follows the style guidelines of this project
- [ ]x I have performed a self-review of my own code
- [ ] I have updated the `CHANGELOG` including this PR

### CODE REVIEW

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/MapboxNavigationNotification.kt [hunk 1]
```diff
@@ -250,20 +251,19 @@ internal class MapboxNavigationNotification : NavigationNotification {
...
     }
 
-    private fun newDistanceText(routeProgress: RouteProgress): Boolean {
-        val formattedDistance = distanceFormatter?.formatDistance(routeProgress.currentLegProgress().currentStepProgress().distanceRemaining()).toString()
-        return currentDistanceText?.let { currentDistance ->
-            currentDistance.toString() != formattedDistance
-        } ?: false
-    }
+    private fun newDistanceText(routeProgress: RouteProgress) =
+            ifNonNull(distanceFormatter, routeProgress.currentLegProgress(), currentDistanceText) { distanceFormatter, currentLegProgress, currentDistanceText ->
```
Comment (Lebedinsky) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/MapboxNavigationNotification.kt]: Should we add newlines when > 2 params?

**Comment by Lebedinsky** [comment]
You have conflicts in

libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/MapboxNavigationNotification.kt
libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/routeprogress/MetricsRouteProgress.java

👀 it please

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/MapboxNavigationNotification.kt (hunk 1 above)
Comment (olegzil) [reply libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/MapboxNavigationNotification.kt]: 👍

File: app/src/main/java/com/mapbox/services/android/navigation/testapp/example/utils/RouteProgressUtils.kt [hunk 2]
```diff
@@ -4,12 +4,11 @@ import android.content.Context
...
 import com.mapbox.services.android.navigation.v5.utils.time.TimeFormatter.formatTime
 import java.util.Calendar
-
-fun RouteProgress.formatArrivalTime(context: Context): String {
-    val time = Calendar.getInstance()
-    val isTwentyFourHourFormat = DateFormat.is24HourFormat(context)
-    // TODO localization 12 /24 time specify from settings
-    return formatTime(time, durationRemaining(), NONE_SPECIFIED, isTwentyFourHourFormat)
-}
+fun RouteProgress.formatArrivalTime(context: Context): String =
```
Comment (RingerJK) [inline app/src/main/java/com/mapbox/services/android/navigation/testapp/example/utils/RouteProgressUtils.kt]: better to move it to `utils.extensions` folder

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/routeprogress/RouteProgress.kt [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+    /**
+     * Get the route the navigation session is currently using. When a reroute occurs and a new
+     * directions route gets obtained, with the next location update this directions route should
+     * reflect the new route. All direction route get passed in through
+     * [com.mapbox.services.android.navigation.v5.navigation.MapboxNavigation.startNavigation].
+     *
+     * @return a [DirectionsRoute] currently being used for the navigation session
+     * @since 0.1.0
+     */
+    fun directionsRoute() = directionsRoute
```
Comment (RingerJK) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/routeprogress/RouteProgress.kt]: already provided by `var directionsRoute`
here and below. don't need additional getters for properties. 
I think we should make fields in constructor `val` and if we'd like to change smt just call `copy()` like `object.copy(propertyA = "newValue")`

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/extensions/NullEx.kt [hunk 4]
```diff
@@ -1,27 +1,27 @@
 package com.mapbox.services.android.navigation.v5.utils.extensions
 
-fun <R1, T> ifNonNull(r1: R1?, func: (R1) -> T): T? =
+inline fun <R1, T> ifNonNull(r1: R1?, func: (R1) -> T): T? =
```
Comment (RingerJK) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/extensions/NullEx.kt]: 👍

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/routeprogress/RouteProgress.kt [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+ *
+ *
+ * The latest route progress object can be obtained through either the [ProgressChangeListener]
+ * or the [com.mapbox.services.android.navigation.v5.milestone.MilestoneEventListener] callbacks.
+ * Note that the route progress object's immutable.
+ *
+ *
+ * @since 0.1.0
+ */
+data class RouteProgress internal constructor(
```
Comment (RingerJK) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/routeprogress/RouteProgress.kt]: why `static builder` was removed? it's easy to use the same approach with koltin. Please take a look at `OfflineTiles` as a sample

File: app/src/main/java/com/mapbox/services/android/navigation/testapp/example/utils/RouteProgressUtils.kt (hunk 2 above)
Comment (olegzil) [reply app/src/main/java/com/mapbox/services/android/navigation/testapp/example/utils/RouteProgressUtils.kt]: I did not create the file. I just modified. The author of the file should move it if that is what's desired

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/routeprogress/RouteProgress.kt (hunk 3 above)
Comment (olegzil) [reply libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/routeprogress/RouteProgress.kt]: I'm not sure I understand. This is a builder. and the method you are referencing is being used other places.

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/routeprogress/RouteProgress.kt (hunk 5 above)
Comment (olegzil) [reply libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/routeprogress/RouteProgress.kt]: As I understand it, you want me to rewrite the builder. To what end?

Comment (RingerJK) [reply libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/routeprogress/RouteProgress.kt]: Let me show a diff.
New approach:
```
new RouteProgress().Builder()
```
old one:
```
RouteProgress.builder()
```
API was broken when we don't have to do it. To keep builder static without need to create an instance directly just move `builder()` fun to `companion object` and mark this function by `@JvmStatic` annotation. Just like in `OfflineTiles` class

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/routeprogress/RouteProgress.kt (hunk 3 above)
Comment (RingerJK) [reply libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/routeprogress/RouteProgress.kt]: this method is not a part of `Builder`, it's a part of data class

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/MapboxNavigationNotification.kt [hunk 6]
```diff
@@ -25,13 +25,18 @@ import com.mapbox.services.android.navigation.v5.navigation.notification.Navigat
...
+import com.mapbox.services.android.navigation.v5.utils.extensions.ifNonNull
 import com.mapbox.services.android.navigation.v5.utils.time.TimeFormatter.formatTime
 import java.util.Calendar
 
 /**
  * This is in charge of creating the persistent navigation session notification and updating it.
  */
 internal class MapboxNavigationNotification : NavigationNotification {
+
+    private val END_NAVIGATION_ACTION = "com.mapbox.intent.action.END_NAVIGATION"
```
Comment (Guardiola31337) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/MapboxNavigationNotification.kt]: Any particular reason why this was refactor out of the `companion object` and removed `const`?

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/MapboxNavigationNotification.kt [hunk 7]
```diff
@@ -393,14 +396,18 @@ internal class MapboxNavigationNotification : NavigationNotification {
...
-            )
+                NavigationConstants.STEP_MANEUVER_TYPE_ROUNDABOUT == maneuverType ||
+                        NavigationConstants.STEP_MANEUVER_TYPE_ROTARY == maneuverType || NavigationConstants.STEP_MANEUVER_MODIFIER_UTURN == maneuverModifier
+                )
     }
 
     private fun onEndNavigationBtnClick() {
-        if (::mapboxNavigation.isInitialized) {
-            mapboxNavigation.stopNavigation()
+        when (::mapboxNavigation.isInitialized) {
```
Comment (Guardiola31337) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/MapboxNavigationNotification.kt]: Isn't 

```kotlin
if (::mapboxNavigation.isInitialized) {
  mapboxNavigation.stopNavigation()
}
```

more readable? Note that `else ->` is empty.

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/route/FasterRouteDetector.java [hunk 8]
```diff
@@ -115,8 +115,7 @@ private boolean validRouteResponse(DirectionsResponse response) {
 
   private boolean validRouteDurationRemaining(RouteProgress routeProgress) {
     // Total route duration remaining in seconds
-    int routeDurationRemaining = (int) routeProgress.durationRemaining();
-    return routeDurationRemaining > VALID_ROUTE_DURATION_REMAINING;
+    return (int)routeProgress.durationRemaining().intValue() > VALID_ROUTE_DURATION_REMAINING;
```
Comment (Guardiola31337) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/route/FasterRouteDetector.java]: Typo

```suggestion
    return (int) routeProgress.durationRemaining().intValue() > VALID_ROUTE_DURATION_REMAINING;
```

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/routeprogress/RouteProgress.kt (hunk 5 above)
Comment (Guardiola31337) [reply libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/routeprogress/RouteProgress.kt]: In any case, whatever we decide we should be consistent across the project.

Comment (olegzil) [reply libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/routeprogress/RouteProgress.kt]: As discussed in our sync, the problem here is this call: toBuilder(). Look at the class MapboxNavigationNotificationTest(), method call checksInstructionTextIsUpdatedIfInstructionTextIsNotInitialized(). In this call buildDefaultTestRouteProgress() returns an object of type RouteProgress, not Builder. From that object the method toBuilder() is called. It must return the correct instance of the previously built RouteProgress object. But we don't have access to the Builder object at this point. We have access only to the RouteProgress object. When the Builder is called to create RouteProgress, it stores itself in the object that it built, thus toBuilder can return the correct instance.

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/MapboxNavigationNotification.kt (hunk 6 above)
Comment (olegzil) [reply libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/MapboxNavigationNotification.kt]: I've answered the question in a prior PR. If values are private to a class there is no need to export them to the outside world via a "companion" object.  In short, a "companion" object exports values and methods to the outside world as public, static values. We should use companions only for that purpose.

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/MapboxNavigationNotification.kt (hunk 7 above)
Comment (olegzil) [reply libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/MapboxNavigationNotification.kt]: Yes, in this case you're correct, but in the general case, "when" will fail if you do not provide handlers for each case tested. This makes "when" safer then if-else. You can forget the "else" clause, but with "when", the compiler will remind you.

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/MapboxNavigationNotification.kt [hunk 9]
```diff
@@ -394,14 +397,18 @@ internal class MapboxNavigationNotification : NavigationNotification {
...
+            true -> {
+                mapboxNavigation.stopNavigation()
+            }
+            else -> {
```
Comment (RingerJK) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/MapboxNavigationNotification.kt]:
```suggestion
            else -> Unit
```

**Pull request merged**